plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/procedure/MigrateProcedure.java

```diff
@@ -0,0 +1,325 @@
+            append.commit();
+
+            transaction.commitTransaction();
+            log.debug("Successfully loaded Iceberg metadata");
+        }
+        catch (Exception e) {
+            catalog.renameTable(session, backupTableName, sourceTableName);
+            throw new TrinoException(ICEBERG_COMMIT_ERROR, "Failed to migrate table", e);
+        }
+    }
+
+    private Map<String, String> icebergTableProperties(String location, Map<String, String> hiveTableProperties, NameMapping nameMapping)
+    {
+        Map<String, String> icebergTableProperties = new HashMap<>();
+
+        icebergTableProperties.putAll(hiveTableProperties);
+        EXCLUDED_PROPERTIES.forEach(icebergTableProperties::remove);
+
+        icebergTableProperties.put("provider", "iceberg");
+        icebergTableProperties.put("migrated", "true");
+        icebergTableProperties.putIfAbsent("location", location);
+        icebergTableProperties.put(DEFAULT_NAME_MAPPING, toJson(nameMapping));
+        icebergTableProperties.put(FORMAT_VERSION, String.valueOf(icebergConfig.getFormatVersion()));
+
+        return ImmutableMap.copyOf(icebergTableProperties);
+    }
+
+    private Schema toIcebergSchema(List<Column> columns)
+    {
+        List<Types.NestedField> icebergColumns = new ArrayList<>();
+        for (Column column : columns) {
+            int index = icebergColumns.size();
+            org.apache.iceberg.types.Type type = toIcebergType(typeManager.getType(column.getType().getTypeSignature()));
+            Types.NestedField field = Types.NestedField.of(index, false, column.getName(), type, column.getComment().orElse(null));
+            icebergColumns.add(field);
+        }
+        org.apache.iceberg.types.Type icebergSchema = Types.StructType.of(icebergColumns);
+        AtomicInteger nextFieldId = new AtomicInteger(1);
+        icebergSchema = TypeUtil.assignFreshIds(icebergSchema, nextFieldId::getAndIncrement);
+        return new Schema(icebergSchema.asStructType().fields());
+    }
+
+    public Map<String, Optional<Partition>> listAllPartitions(HiveMetastore metastore, io.trino.plugin.hive.metastore.Table table)
+    {
+        List<String> partitionNames = table.getPartitionColumns().stream().map(Column::getName).collect(toImmutableList());
+        Optional<List<String>> partitions = metastore.getPartitionNamesByFilter(table.getDatabaseName(), table.getTableName(), partitionNames, TupleDomain.all());
+        if (partitions.isEmpty()) {
+            return ImmutableMap.of();
+        }
+        return metastore.getPartitionsByNames(table, partitions.get());
+    }
+
+    private List<DataFile> buildDataFiles(ConnectorSession session, HiveStorageFormat format, String location, PartitionSpec partitionSpec, String partitionKey)
+            throws IOException
+    {
+        // TODO: Introduce parallelism
+        TrinoFileSystem fileSystem = fileSystemFactory.create(session);
+        FileIterator files = fileSystem.listFiles(location);
```

Comment:
Can it be that BackgroundHiveSplitLoader would discover different set of files than we do?

For example, this code doesn't take `hive.recursive-directories` into account.

Reply:
`recursive` argument in `HdfsFileSystem#listFiles` is always true. 
https://github.com/trinodb/trino/blob/1759cb5e69c9ec3c864ef22e8c961063707de02e/lib/trino-filesystem/src/main/java/io/trino/filesystem/hdfs/HdfsFileSystem.java#L115

Reply:
is this implying that `hive.recursive-directories` config no longer means anything?

Reply:
That's a Hive connector property, which doesn't exist in Iceberg. My concern is that we default this to `false` in the Hive connector, so this would produce different results. Athena sets it to `true`, so maybe this is fine and we should revisit the Trino behavior.

What does the Iceberg Spark code do when migrating?

Reply:
It seems Spark doesn't migrate nested directories as far as I confirmed the behavior using CLI and the relevant code (https://github.com/apache/iceberg/blob/ab6391767e75357a9a12691513d9f83a4180f7e4/data/src/main/java/org/apache/iceberg/data/TableMigrationUtil.java#L106). 

Can we add another boolean argument (e.g. `RECURSIVE_DIRECTORIES`) to this procedure?

Reply:
@electrum @findepi Added another optional argument to the procedure and rename the existing arguments for consistency with `register_table` procedure.  Could you take another look when you have time?